lib/puppet/application/resource.rb

```diff
@@ -135,6 +136,18 @@ def help
 
   def main
     env = Puppet.lookup(:environments).get(Puppet[:environment])
+
+    # If the specified environment does not exist locally, fall back to the
+    # default (production) environment
+    unless env
+      Puppet.debug("Specified environment '#{Puppet[:environment]}' does not exist on the filesystem, defaulting to 'production'")
+      Puppet[:environment] = :production
+      basemodulepath = Puppet::Node::Environment.split_path(Puppet[:basemodulepath])
+      modulepath = Puppet[:modulepath]
+      modulepath = (modulepath.nil? || '' == modulepath) ? basemodulepath : Puppet::Node::Environment.split_path(modulepath)
+      env = Puppet::Node::Environment.create(Puppet[:environment], modulepath, Puppet::Node::Environment::NO_MANIFEST)
```

Comment:
@joshcooper is it too late to create the default environment here? I based the code on https://github.com/puppetlabs/puppet/blob/7.8.0/lib/puppet.rb#L230-L237

Reply:
I think it's ok. The only time type/providers would be loaded before would be when applying the settings catalog. But the user/group/file types are builtin, so don't depend on an environment.